src/utils/fixedptc.h

```diff
@@ -194,66 +197,108 @@ static inline fixedpt fixedpt_div(fixedpt A, fixedpt B) {
  * of decimal digits will be used (2 for 32-bit fixedpt width, 10 for
  * 64-bit fixedpt width); If set to -2, "all" of the digits will
  * be returned, meaning there will be invalid, bogus digits outside the
- * specified precisions.
+ * specified precisions. If set to -3, all the zero digits will be trimmed
+ * from the end of the decimal digits.
+ *
+ * Returns the number of bytes written to `str`.
  */
-static inline void fixedpt_str(fixedpt A, char *str, int max_dec) {
+static inline size_t fixedpt_str(fixedpt A, char *str, size_t bufsize, int max_dec) {
     // boy this function sure is safe with pointers :(
-    int ndec = 0, slen = 0;
-    char tmp[12] = {0};
+    size_t nint = 0;  // number of integer digits
+    size_t nfrac = 0; // number of fractional digits
+    size_t slen = 0;  // string length used
+
+    // accumulator for integer digits, can hold the max number of digits for FIXEDPT_BITS
+    char acc[FIXEDPT_WCHARS] = {0};
     fixedptud fr, ip;
+    fixedptud magnitude = (fixedptud)A;
```

Comment:
i find the signed to unsigned cast of a different bitwidth a bit scary, could you make the sign extension explicit:
```suggestion
    fixedptud magnitude = (fixedptud)(fixedptd)A;
```
or, if you don't want the value to be sign extended:
```c
    fixedptud magnitude = (fixedptud)(fixedptu)A;
```